Use hashed values for method names in the CompiledContainer

holtkamp opened the pull request:
As suggested in https://github.com/PHP-DI/PHP-DI/issues/604

mnapoli commented on src/Compiler/Compiler.php:
```diff
@@ -260,8 +272,8 @@ public function compileValue($value) : string
         if ($value instanceof Definition) {
-            // Give it an arbitrary unique name
-            $subEntryName = uniqid('SubEntry');
+            $subEntryName = $this->getHashedValue('SubEntry', $value->getName() . $value);
```
I am cautious about this part: nested definitions don't have names, so `getName()` will probably be null here.

And casting `$value` to string dumps the debug output of the definition. That output is not unique ([for example](https://github.com/PHP-DI/PHP-DI/blob/fd7bde0fbd25b7e4ca913907d274b9a71278e347/src/Definition/FactoryDefinition.php#L77)).

I'm not sure what to do here that would not require bigger changes 🤔

holtkamp replied:
yeah, was getting that concern this morning too. But then again: SubEntries get the value of there "parent" entry appended as well since the `$subEntryName` is handed over to `compileDefinition()`... and the parent must be unique, right?

Maybe some more exhaustive testing could be added. For what its worth: adopted this fork in production without problems

mnapoli replied:
I'm not sure I understand what you mean. Here the `$subEntryName` is used to create a method in the compiled container class, it needs to be unique in itself?

Maybe what we could do is propagate the parent entry names down the nested definitions somehow (that way we can prefix those). But I'm afraid that means adding extra parameters everywhere.

holtkamp replied:
>I'm not sure I understand what you mean.

Ah, yeah sorry, that did not make sense. 

So theoretically a collision could occur in case multiple Definitions exist which have:
- the name "<nested definition>", for example for a
  - `DI\Definition\Reference`
  - `DI\Definition\ArrayDefinition`
  - ... ?? more possible?
- the same value

Let's add a (failing) test for that first

mnapoli replied:
That is interesting! FYI I won't have time to look further into that in the next days. We could add a new method on definitions (e.g. `getHash()`) if necessary.

Also I think we need to check explicitly for collisions with closures, because `__toString()` is very poor right now so I think there may be collisions.

holtkamp replied:
>I won't have time to look further into that in the next days.

ok, standing by, for now this approach resolves our issue. 

>We could add a new method on definitions (e.g. getHash()) if necessary.

Indeed, but then a Definition needs to be aware that it is a SubEntry as well?

>Also I think we need to check explicitly for collisions with closures, because __toString() is very poor right now so I think there may be collisions.

True, my knowledge of the internals is too limited to set up proper tests...

Step-by-step we'll get there 😄

mnapoli commented on tests/IntegrationTest/CompiledContainerTest.php:
```diff
@@ -9,6 +9,8 @@
 use DI\Definition\Exception\InvalidDefinition;
 use function DI\get;
+use DI\Test\IntegrationTest\Definitions\NestedDefinitionsTest\AllKindsOfInjections;
+use DI\Test\IntegrationTest\Definitions\NestedDefinitionsTest\Autowireable;
```
Could you copy those classes in this file (like [done here](https://github.com/PHP-DI/PHP-DI/blob/79712dc23ae1a86824dde1ee659ddc45449d275c/tests/IntegrationTest/Definitions/NestedDefinitionsTest.php#L238-L265))? I refactored the tests with v6 so that most integration tests are isolated from each other, that is much easier to maintain.

holtkamp replied:
sure, was in a hurry

holtkamp commented on src/Compiler/Compiler.php:
```diff
@@ -141,15 +141,27 @@ public function compile(
+     */
+    private function getHashedValue(string $prefix, string $value): string
+    {
+        return sprintf('%s%s', $prefix, md5($value));
```
Haha, guess that is my personal quirk of loving `sprintf()`

holtkamp commented:
Thanks for the feedback, will try to process it soon!

mnapoli commented on src/Compiler/Compiler.php:
```diff
@@ -141,15 +141,27 @@ public function compile(
     {
-        // Generate a unique method name
-        $methodName = str_replace('.', '', uniqid('get', true));
+        $methodName = $this->getHashedValue('get', $entryName . $definition);
```
Oh I missed the concatenation to the `$definition` cast to string. I don't think that is necessary as entry names are unique (and that is expensive since `$definition` is an object).

holtkamp replied:
confirmed, appending the string representation of the Definition is only required for the subEntries

holtkamp commented:
@mnapoli processed your feedback and pushed the changes. 

Can you think of some additional Definitions to add to the tests? As indicated: in a real-life situation no collisions occurred, so it seems it all works out. But it is probably better to be safe than sorry so the more we test, the better 😄

holtkamp commented:
I added some tests. For now I conclude that when a collision occurs in the entryName resulting from the hash, the value of the Definition should be the same. 

Added a check for that. In case they ARE the same (which they should), then this can be considered an optimization: the amount of functions with different names, but equal body (so functionally nothing changes), is reduced.

holtkamp commented:
FYI: rebased changes on master branch

holtkamp commented:
FYI: rebased changes on master branch

@mnapoli I know you are busy, but do you, or someone else, have ideas on getting this functionality merged?

mnapoli commented:
@holtkamp really sorry about the delay. Regarding the last discussion left open in the diff it seems to me you've covered it with a check (an exception is thrown). Is that right? That seems like a good call, at least a clear exception is thrown. (just want to make sure before I merge, it's been some time 😄)

By the way have you been able to use this branch in production?

holtkamp commented:
@mnapoli indeed, a `InvalidDefinition` Exception is thrown in case "equal hashes are involved, which do not point to the same method name". 

>By the way have you been able to use this branch in production?

yes, since mid-june 2018, works properly / no strange side-effects...

elie29 commented on src/Compiler/Compiler.php:
```diff
@@ -157,15 +157,33 @@ public function compile(
+     */
+    private function getHashedValue(string $prefix, string $value): string
+    {
+        return $prefix . md5($value);
```
@holtkamp IMHO, I think we should not use md5 but sha1 or hash function

holtkamp replied:
@elie29 ok, but why would that have your preference? Due to the chance of collision? That change is really small..

https://stackoverflow.com/questions/201705/how-many-random-elements-before-md5-produces-collisions

Or is it for performance reasons?

mnapoli replied:
A quick search shows that md5 is faster than sha1 so that sounds better to use md5.

elie29 replied:
@holtkamp Just due to collision

mnapoli replied:
I really don't think we have to worry about that given the size of the sample.

mnapoli commented:
Thank you!

I've tagged a new release: https://github.com/PHP-DI/PHP-DI/releases/tag/6.0.7

holtkamp commented:
:shipit:
